### PR TITLE
Cayley Graph Notebook

### PR DESCRIPTION
**Pull request by joshp112358**

### CODE REVIEW

File: libsemigroups_cppyy/froidure_pin.py [hunk 1]
```diff
@@ -22,4 +21,10 @@ def FroidurePin(*args):
     froidure_pin_type = cppyy.gbl.libsemigroups.FroidurePin(types.pop())
     detail.unwrap(froidure_pin_type, froidure_pin_type.factorisation, lambda x:
             list(x))
+    detail.unwrap(froidure_pin_type, froidure_pin_type.left, lambda x:
```
Comment (james-d-mitchell) [inline libsemigroups_cppyy/froidure_pin.py]: You can use `detail.unwrap_int` for each of these three, which auto fills in the lambda function which is the final parameter.

File: libsemigroups_cppyy/froidure_pin.py [hunk 2]
```diff
@@ -1,14 +1,13 @@
...
     https://libsemigroups.readthedocs.io/en/latest/_generated/libsemigroups__froidurepin.html
-
 for further details.
 """
 
 import cppyy
 import libsemigroups_cppyy.detail as detail
-
+import numpy as np
+from array import array
```
Comment (james-d-mitchell) [inline libsemigroups_cppyy/froidure_pin.py]: Since there's no code in this file requiring `numpy` or `array`, please remove these two lines.

File: libsemigroups_cppyy/__init__.py [hunk 3]
```diff
@@ -44,6 +44,7 @@
 cppyy.include("libsemigroups/knuth-bendix.hpp")
 cppyy.include("libsemigroups/schreier-sims.hpp")
 cppyy.include("libsemigroups/report.hpp")
+cppyy.include("libsemigroups/cong-pair.hpp")
```
Comment (james-d-mitchell) [inline libsemigroups_cppyy/__init__.py]: This change also seems to be unrelated to the other changes in the PR.

**Review by james-d-mitchell — CHANGES_REQUESTED** [review]
This is a great start, I made some minor comments, but I think we might have misunderstood each other about the nature of this PR. I suggest adding the Cayley graph function that is contained in the jupyter notebook to the python code in libsemigroups_cppyy (i.e. take it out of the ipython notebook and put it into the froidure-pin.py file).

File: libsemigroups_cppyy/__init__.py [hunk 4]
```diff
@@ -17,6 +17,7 @@
 
 import cppyy
 import os
+import networkx
```
Comment (james-d-mitchell) [inline libsemigroups_cppyy/__init__.py]: We're getting there. I don't think this line needs to be here, please delete.

File: libsemigroups_cppyy/froidure_pin.py [hunk 5]
```diff
@@ -22,4 +20,25 @@ def FroidurePin(*args):
...
     return froidure_pin_type(gens)
+
+def right_cayley(S):
+    edges = []
+    for i in range(S.size()):
+        for j in range(S.nr_generators()):
+            edges.append((str(S.factorisation(S[i])),str(S.factorisation(S.right(i,j)))))
+    G=networkx.DiGraph()
+    G.add_edges_from(edges)
+    return(networkx.draw(G, with_labels = True))
```
Comment (james-d-mitchell) [inline libsemigroups_cppyy/froidure_pin.py]: I think this should just return G, not the picture of G. Also might it not be better to create G at the start of the function, and then add edges to it in the nested loops, rather than creating the list edges, which is almost immediately thrown away? I'm not sure about the capabilities of networkx, perhaps this isn't possible.

File: setup.py [hunk 6]
```diff
@@ -10,7 +10,7 @@
       author='James D. Mitchell, Nicolas Thiéry',
       author_email='jdm3@st-andrews.ac.uk, Nicolas.Thiery@u-psud.fr',
       license='GPL3',
-      install_requires=['cppyy'],
+      install_requires=['cppyy','matplotlib','networkx'],
```
Comment (james-d-mitchell) [inline setup.py]: I also don't see why matplotlib is required, please remove this too? Thanks!

**Review by james-d-mitchell — CHANGES_REQUESTED** [review]
We're getting there, please update according to the comments I've made, and also remove the `.DS_Store` file from the PR, this is a Mac thing, and isn't part of your PR.

File: setup.py (hunk 6 above)
Comment (joshp112358) [reply setup.py]: We needed matplotlib to show the graph but now that we are just return a networkx graph object, it's fine.

**Comment by joshp112358** [comment]
Added in pyvis to notebook
